### PR TITLE
study(color): price P9 item 2's other half — building in Q16, not just inverting (#4043)

### PR DESCRIPTION
**Pull request by zackees**
## #4275 measured half the precondition; this is the other half

P9's survey calls item 2 — "fixed-point `EmitterProfile` + bind-time derivation" — the piece "genuinely blocked on nothing but effort". #4275 then measured its precondition: a Q16 3×3 inverse can replace the float one, at **0.0064 dE2000** against A1's 0.5.

But that measurement kept the **forward** matrix float. It quantised a matrix float had already built from float chromaticities — which is not what item 2 needs. `EmitterProfile` stores `float xy_r[2]`, and a float-free bind path has to hold those as Q16 and run `xyY_to_XYZ` in Q16 too.

There was a specific reason to doubt that half. `xyY_to_XYZ` is `X = Y·x/y` — a **division by `y`**, and BT.2020's blue sits at `y = 0.046`. One Q16 step there is a relative perturbation of `2.5e-04` on a divisor, landing on the column an inverse is least able to absorb.

## It holds — and the cost is 17× the inverse-only figure

| primaries | chroma q16 err | coef ULP | drive err | worst dE2000 |
|---|---:|---:|---:|---:|
| srgb | 4.07e-05 | 1 | 2.14e-04 | 0.0172 |
| **bt2020** | **1.14e-04** | 2 | 3.66e-04 | **0.1085** |
| display_p3 | 4.07e-05 | 1 | 7.63e-05 | 0.0037 |
| narrow | 1.53e-05 | 34 | 9.16e-05 | 0.0025 |

Worst **0.1085** against A1's 0.5, of which the derivation has roughly 0.35 once the gamut mapper's 0.15 is accounted for.

So **item 2 is feasible** — but the profile quantisation, not the inverse, is where its error budget goes: 0.109 against #4275's 0.0064 on the same corpus. Anyone doing item 2 should size it against 0.109, not 0.0064.

## The predicted mechanism is the one that shows up

BT.2020 carries both the worst chromaticity residual and the worst dE2000, and it is the profile with the smallest `y`. `narrow`, which is near-degenerate, takes **34 coefficient ULP and still only 0.0025 dE2000** — conditioning moves the coefficients, the small divisor moves the colour, and they are not the same failure. Worth stating because #4275's degeneracy walk found the opposite ordering, and it is easy to assume one figure predicts the other.

## Checked rather than assumed

**Undersampling.** The figure is stable from a 9-step drive sweep upward — 0.1085 at 9, 0.1093 at 17, 0.1000 at 25 — and only a 5-step sweep understates it, at 0.0423. The default is 9.

**Vacuity.** The most important case in the test file checks the two arms actually differ. A study whose candidate silently converged on its baseline reports "inside budget" having measured nothing.

| mutation | result |
|---|---|
| build the candidate from float chromaticities (arms converge) | headline case fails |
| remove the zero-`y` guard | degenerate case fails |

`bash lint` clean; 6 cases pass.

## What this does not do

It does not convert anything. `buildRgbSolveMatrixQ16` still builds and inverts in float, and `EmitterProfile` still stores floats — that is item 2, and it stays cross-cutting with P2 because the legacy RGBW stack consumes the same type. This says the effort is worth spending and what it has to fit inside.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MLhWkMfzLjrnLTDMBE6Fj9

## Summary by CodeRabbit

- **New Tools**
  - Added a fixed-point color profile study comparing quantized Q16 calculations with the existing floating-point approach.
  - Reports color accuracy, drive error, coefficient precision, and chromaticity quantization results across supported primary profiles.
  - Returns a failure status when measured color error exceeds the defined accuracy budget.

- **Tests**
  - Added coverage for quantization behavior, accuracy thresholds, edge cases, and successful study execution.

### CODE REVIEW

File: ci/tests/test_color_fixed_profile_study.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    emitter_matrix_q16,
+    main,
+    quantization_residual,
+)
+
+
+class TestQuantizedProfileStudy(unittest.TestCase):
+    """The measurement, and the reasons to believe it."""
+
+    def test_the_two_arms_actually_differ(self) -> None:
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_color_fixed_profile_study.py:34]: Declining this one, with the measurement behind it.

Annotating `self` is not this repo's convention. Across `ci/tests/*.py` the count is **43 test methods with `self:` against 630 without** — so the existing corpus is 94% the other way, and `bash lint` (ruff, `ty`, and the repo's own checkers) passes on both forms.

The guideline the finding quotes — "all function parameters MUST have explicit type annotations" — is real, but `self` is not a parameter a caller supplies and its type is fixed by the class. Applying it here would make these files the odd ones out rather than bring them into line.

Happy to be overruled if the intent is that all 630 get changed; that would be a repo-wide sweep with its own PR rather than something to start in this one.

File: ci/tests/test_color_fixed_profile_study.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+        worst = 0.0
+        for name, primaries in corpus_primaries():
+            measured = compare_profile_quantization(name, primaries, 9)
+            if measured.delta_e > worst:
+                worst = measured.delta_e
+        self.assertLess(worst, 0.35)
+        # And the measured value, so a regression that stays inside the
+        # budget still fails rather than sliding by.
+        self.assertLess(worst, 0.15)
```
[the remark this answers is not in the document]
Comment (zackees) [reply ci/tests/test_color_fixed_profile_study.py]: Right, and fixed. `assertLess(worst, 0.15)` admitted 0.149 — 37% worse than the 0.1085 the issue comment reports, passing silently. A study whose test allows a materially different number is not pinning the result it published.

It is now a band of ±5% around the measurement: `0.1031 < worst < 0.1139`. That is far outside anything the arithmetic can drift by — both arms are deterministic integer paths over a fixed corpus and a fixed sweep, so the figure is reproducible to the digit — and far inside the 38% headroom to the derivation's 0.35 share, so it does not turn a budget question into a tolerance question.

Demonstrated rather than argued. Holding the profile one, two or three bits coarser than Q16 — a plausible design choice for item 2, not a contrived edit:

| profile precision | worst dE2000 | old `< 0.15` | new band |
|---|---:|---|---|
| Q15 | 0.1021 | passes | **fails** |
| Q14 | 0.1024 | passes | **fails** |
| Q13 | 0.1160 | passes | **fails** |

The lower bound also subsumes the "not zero" check it replaces: a candidate that silently converged on its baseline reports 0 and fails the band from below, which is the same protection expressed as one range rather than two unrelated comparisons.
